parallel-install/pkg/preinstaller/preinstaller.go

```diff
@@ -0,0 +1,127 @@
+package preinstaller
+
+import (
+	"fmt"
+	"github.com/avast/retry-go"
+	"github.com/kyma-incubator/hydroform/parallel-install/pkg/config"
+	"io/ioutil"
+	"k8s.io/client-go/dynamic"
+	"os"
+)
+
+// PreInstaller prepares k8s cluster for Kyma installation.
+type PreInstaller struct {
+	applier       ResourceApplier
+	cfg           config.Config
+	dynamicClient dynamic.Interface
+	retryOptions  []retry.Option
+}
+
+// NewPreInstaller creates a new instance of PreInstaller.
+func NewPreInstaller(applier ResourceApplier, cfg config.Config, dynamicClient dynamic.Interface, retryOptions []retry.Option) *PreInstaller {
+	return &PreInstaller{
+		applier:       applier,
+		cfg:           cfg,
+		dynamicClient: dynamicClient,
+		retryOptions:  retryOptions,
+	}
+}
+
+// InstallCRDs on a k8s cluster.
+// Returns Output containing results of installation.
+func (i *PreInstaller) InstallCRDs() (Output, error) {
+	resource := resourceType{
+		name: "crds",
+	}
+
+	output, err := i.apply(resource)
+	if err != nil {
+		return Output{}, err
+	}
+
+	return output, nil
+}
+
+// CreateNamespaces in a k8s cluster.
+// Returns Output containing results of installation.
+func (i *PreInstaller) CreateNamespaces() (Output, error) {
+	resource := resourceType{
+		name: "namespaces",
+	}
+
+	output, err := i.apply(resource)
+	if err != nil {
+		return Output{}, err
+	}
+
+	return output, nil
+}
+
+func (i *PreInstaller) apply(resourceType resourceType) (o Output, err error) {
```

Comment:
I think this function does a bit too much. It is responsible for **finding** resources and for **applying** them.
How about splitting this functionality?
I'd suggest to have two functions:
findResources that does the finding/directory reads and returns a list of objects with fields like {resourceType, component, fileName, path} (only necessary data)
applyResources that is processing the list produced by findResources

I would also split this nested for loop into two private functions each having a single for loop. But I will not insist on that, if the code is simple, a one-level nesting is still OK.